pkg/networkserver/redis/registry.go

```diff
@@ -480,305 +536,110 @@ func (r *DeviceRegistry) RangeByUplinkMatches(ctx context.Context, up *ttnpb.Upl
+		pendingSessionScoresCmd  *redis.ZSliceCmd
+		pendingSessionMappingCmd *redis.StringStringMapCmd
+	)
+	if _, err := r.Redis.TxPipelined(ctx, func(p redis.Pipeliner) error {
+		currentSessionScoresCmd = ZRangeArgsWithScores(ctx, p, addrKeyCurrent)
+		currentSessionMappingCmd = p.HGetAll(ctx, fieldKeyCurrent)
+		if !ackFlag {
+			pendingSessionScoresCmd = ZRangeArgsWithScores(ctx, p, addrKeyPending)
+			pendingSessionMappingCmd = p.HGetAll(ctx, fieldKeyPending)
 		}
+		return nil
+	}); err != nil {
+		return ttnredis.ConvertError(err)
 	}
 
-	trace.Log(ctx, "ns:redis", "run deviceMatchScript")
-	vs, err := ttnredis.RunInterfaceSliceScript(ctx, r.Redis, deviceMatchScript, matchKeys, lsb, cacheTTL.Milliseconds()).Result()
+	var (
+		currentSessionSet []sessionEntry
+		pendingSessionSet []sessionEntry
+		err               error
+	)
+	currentSessionSet, err = buildSortedSessionSet(currentSessionScoresCmd, currentSessionMappingCmd, true, lsb)
 	if err != nil {
-		if err == redis.Nil {
-			return errNoUplinkMatch.WithCause(errNoCandidates)
-		}
-		return ttnredis.ConvertError(err)
+		return err
 	}
-	if len(vs) < 1 {
-		panic("empty match script return value")
+	if !ackFlag {
+		pendingSessionSet, err = buildSortedSessionSet(pendingSessionScoresCmd, pendingSessionMappingCmd, false, 0)
+		if err != nil {
+			return err
+		}
 	}
-	matchType, ok := vs[0].(string)
-	if !ok {
-		panic(fmt.Sprintf("expected first match script return value element to be a string, got '%v'(%T)", vs[0], vs[0]))
+
+	fillContext := func(ctx context.Context, uid string) (context.Context, *ttnpb.EndDeviceIdentifiers, error) {
+		ids, err := unique.ToDeviceID(uid)
+		if err != nil {
+			log.FromContext(ctx).WithError(err).Error("Failed to parse UID as device identifiers")
+			return ctx, nil, errDatabaseCorruption.WithCause(err)
+		}
+		return ctx, &ids, nil
 	}
-	processResult := func(ctx context.Context, s string) error {
-		defer trace.StartRegion(ctx, "process result").End()
 
-		if s == string(noUplinkMatchMarker) {
-			return errNoUplinkMatch.WithCause(errNoMatchMarker)
+	for _, session := range currentSessionSet {
+		ctx, ids, err := fillContext(ctx, session.UID)
+		if err != nil {
+			return err
 		}
-		ctx = log.NewContextWithField(ctx, "match_key", matchResultKey)
-		res := &UplinkMatchResult{}
-		if err := msgpack.Unmarshal([]byte(s), res); err != nil {
-			log.FromContext(ctx).WithError(err).Error("Failed to unmarshal match result")
-			return errDatabaseCorruption.WithCause(err)
+		if ids == nil {
+			continue
 		}
-		ctx = log.NewContextWithField(ctx, "device_uid", res.UID)
-		ids, err := unique.ToDeviceID(res.UID)
+
+		ses := &UplinkMatchSession{}
+		err = msgpack.Unmarshal([]byte(session.Session), ses)
 		if err != nil {
-			log.FromContext(ctx).WithError(err).Error("Failed to parse match result UID as device identifiers")
-			return errDatabaseCorruption.WithCause(err)
+			continue
+		}
+
+		if uint16(ses.LastFCnt) > lsb {
```

Comment:
This condition comes from the 'greater-than' scan script (negated, almost).
https://github.com/TheThingsNetwork/lorawan-stack/blob/e0b838c1f4ec534cb4d1b97190f0da72b22a0e01/pkg/networkserver/redis/lua/deviceMatchScanGT.lua#L25-L27